### PR TITLE
Enable Structured Losses

### PR DESCRIPTION
**Pull request by nicolaspi**
### Description

This pull request enhances loss configuration flexibility by allowing the use of a nested structure that mirrors the output structure to map outputs to loss functions. This enables training with structured losses that can be applied to any level of the output hierarchy such as individual tensors or nested structures of tensors.

Key features:

1. Flexibility to apply losses to both individual tensors (leaf nodes) and nested structures of tensors (intermediate nodes) in the output.
2. The entire subtree at a given node is passed to the corresponding loss function, allowing for custom loss calculations on structured data.

### Usage Example

```python
import keras
import numpy as np
from keras.src import tree

# Generate some dummy data
np.random.seed(0)
data_shape = (10, 2)
X = np.random.rand(*data_shape)
y1 = np.random.rand(*data_shape)
y2 = np.random.rand(*data_shape)
y3 = np.random.rand(*data_shape)
Y = {'a': y1, 'b': {'c': (y1, y2), 'd': y3}}


# Loss function working on arbitrary pytrees
def loss_fn(y_true, y_pred):
  flat_y_pred, flat_y_true = tree.flatten(y_pred), tree.flatten(y_true)
  diff = 0
  for y_p, y_t in zip(flat_y_pred, flat_y_true):
    diff += keras.losses.mean_absolute_error(y_t, y_p)
  return diff


# Define the model
inputs = keras.Input(shape=data_shape[1:])
x = keras.layers.Dense(64, activation='relu')(inputs)
x = keras.layers.Dense(32, activation='relu')(x)
y1 = keras.layers.Dense(data_shape[-1], name='y1')(x)
y2 = keras.layers.Dense(data_shape[-1], name='y2')(x)
y3 = keras.layers.Dense(data_shape[-1], name='y3')(x)

model = keras.Model(inputs=inputs,
                    outputs={'a': y1, 'b': {'c': (y1, y2), 'd': y3}})

# The following losses configuration applies the loss to the paths:
# 'a', 'b', 'b/c', 'b/c/0', 'b/d'
loss = {
         'a': loss_fn,
         'b': [loss_fn, 
                {
                  'c': [loss_fn, (loss_fn, None)],
                  'd': loss_fn
                 }
              ], 
        }

model.compile(optimizer='adam', loss=loss)

model.fit(X, Y, epochs=4, batch_size=2, verbose=1)

```

Example output:
```
Epoch 1/4
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - a_loss: 0.5786 - b/c/y1_loss: 0.5786 - b/c_loss: 0.9755 - b/d_loss: 0.5344 - b_loss: 1.5099 - loss: 4.1770  
Epoch 2/4
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - a_loss: 0.5413 - b/c/y1_loss: 0.5413 - b/c_loss: 0.9610 - b/d_loss: 0.3915 - b_loss: 1.3525 - loss: 3.7876
Epoch 3/4
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - a_loss: 0.5184 - b/c/y1_loss: 0.5184 - b/c_loss: 0.8692 - b/d_loss: 0.2813 - b_loss: 1.1505 - loss: 3.3379
Epoch 4/4
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - a_loss: 0.4610 - b/c/y1_loss: 0.4610 - b/c_loss: 0.7571 - b/d_loss: 0.2707 - b_loss: 1.0279 - loss: 2.9779
```

### CODE REVIEW

File: keras/src/losses/losses.py
Comment (fchollet) [inline keras/src/losses/losses.py:9]: Keras is designed to not need optree (as we found it cannot be run in some environments...). If you need a new optree function it should be added in `keras.tree` and made to work with dm-tree as well

File: keras/src/losses/losses.py
Comment (fchollet) [inline keras/src/losses/losses.py:27]: Why is this necessary?

File: keras/src/models/model_test.py
Comment (fchollet) [inline keras/src/models/model_test.py:83]: The use of e.g. "name/1" opens the door to collisions (e.g. if a layer has a number as its name). I would recommend using "name/other_name:1", etc.

Even better would be to drop paths and just pass a nested structure that matches the model's outputs.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras/src/models/model_test.py
Comment (nicolaspi) [reply keras/src/models/model_test.py:83]: > The use of e.g. "name/1" opens the door to collisions (e.g. if a layer has a number as its name). I would recommend using "name/other_name:1", etc.

I agree, this is done on purpose here for the unit test. There is no prevention for users to use `/` in the names of their outputs. The test is meant to explore the edge cases where the key naming might collide with the path separator.

> Even better would be to drop paths and just pass a nested structure that matches the model's outputs.

I'm not sure I follow your suggestion. Could you elaborate? To clarify, this is the model's output definition, not the loss configuration. The proposed approach allows users to specify losses for nested outputs using a flat dictionary with path-like keys, which provides flexibility in defining losses for complex output structures. Are you proposing an alternative method for specifying losses?

File: keras/src/losses/losses.py
Comment (nicolaspi) [reply keras/src/losses/losses.py:27]: Do you mean why not simply use `tree_map`?
A simple map would produce a single pytree, but need two for `y_true` and `y_pred`.

File: keras/src/losses/losses.py
Comment (nicolaspi) [reply keras/src/losses/losses.py:9]: That's unfortunate. I make heavy use of optree's advanced features. This limitation will require some rethinking of the current implementation.

File: keras/src/models/model_test.py
Comment (fchollet) [reply keras/src/models/model_test.py:83]: Yes, why not specify nested losses with a nested struct that mirrors the structure of the model's outputs?

Comment (nicolaspi) [reply keras/src/models/model_test.py:83]: Also, the approach using dictionaries with path-like keys offers backward compatibility. While using nested structures for loss specification could lead to compatibility issues with legacy code.
For example, we can have:
```python
output = [[y1, y2], y3]
loss = [loss1, [loss2, loss3]]
```
This configuration is valid with the legacy API because both loss and output are flattened. However, it would be invalid for a strict struct-like specification or lead to ambiguity at best.

Comment (fchollet) [reply keras/src/models/model_test.py:83]: > Use None as a placeholder for 'no loss', e.g., loss = [None, loss_subtree2]

This is already the way it works today! e.g. if you have a model with 2 outputs expressed as a tuple, you can specify a loss via a tuple such as `(None, fn)`.

> This configuration is valid with the legacy API because both loss and output are flattened. However, it would be invalid for a strict struct-like specification or lead to ambiguity at best.

Breaking niche edge cases like this one should be ok I think.

My main concern here is dev UX -- string paths are a rather unusual way to refer to entries in a complex struct, but mirroring the same struct should be more intuitive. It's also more consistent with the way we handle non-nested structs today.

File: keras/src/losses/losses.py
Comment (nicolaspi) [reply keras/src/losses/losses.py]: I removed the optree dependency.

File: keras/src/losses/losses.py
Comment (nicolaspi) [reply keras/src/losses/losses.py]: I removed  `squeeze_or_expand_to_same_rank` altogether, as far as I know there was a `tensorboard` test falling that I fixed.

**Comment by nicolaspi** [comment]
> There was an early attempt for nested/structural loss design: #19879

@james77777778 do you think your [comment](https://github.com/keras-team/keras/pull/19879#discussion_r1646825312) still applies if we constrain `y_true` and `y_pred` to have the same structure?

@fchollet can you review my changes? I switched to a struct-like approach.

Thanks

File: keras/src/callbacks/tensorboard_test.py
Comment (fchollet) [inline keras/src/callbacks/tensorboard_test.py:424]: Why is this change necessary?

File: keras/src/losses/losses.py
Comment (fchollet) [inline keras/src/losses/losses.py:26]: Remove this is likely to break some cases -- we should make sure we keep per-output squeezing/expansion

File: keras/src/callbacks/tensorboard_test.py
Comment (nicolaspi) [reply keras/src/callbacks/tensorboard_test.py:424]: This is the only unit test breaking after removal of the `y_true, y_pred = squeeze_or_expand_to_same_rank(y_true, y_pred)` line related to your other [comment](https://github.com/keras-team/keras/pull/20358#discussion_r1803590378)

File: keras/src/losses/losses.py
Comment (nicolaspi) [reply keras/src/losses/losses.py:26]: So far only one _minor_ unit test was affected cf your [comment](https://github.com/keras-team/keras/pull/20358#discussion_r1803589503).
I'll put it back.

File: keras/src/callbacks/tensorboard_test.py
Comment (nicolaspi) [reply keras/src/callbacks/tensorboard_test.py]: Reverted

File: keras/src/losses/losses.py
Comment (nicolaspi) [reply keras/src/losses/losses.py:26]: Put it back.

**Comment by nicolaspi** [comment]
@james77777778  Since you have worked on this issue and authored a big chunk of the code I am working on, your comments on this PR are very welcome!

**Review by fchollet — COMMENTED** [review]
LGTM, thank you! Will also wait for comments from Hongyu.

File: keras/src/trainers/compile_utils.py
Comment (james77777778) [inline keras/src/trainers/compile_utils.py:489]: Will this causes a collision in `history` when recording individual losses?

File: keras/src/models/model_test.py
Comment (james77777778) [inline keras/src/models/model_test.py:1024]: `type` should be a preserved word, and I'm unsure why we need `type` and `anti_type` in this test.

File: keras/src/models/model_test.py
Comment (james77777778) [inline keras/src/models/model_test.py:1135]: Same here

File: keras/src/trainers/compile_utils.py
Comment (james77777778) [inline keras/src/trainers/compile_utils.py:412]: Is this being used? If so, why do we need it?

File: keras/src/trainers/compile_utils.py
Comment (james77777778) [inline keras/src/trainers/compile_utils.py:610]: Should this be `elif`?

**Review by james77777778 — COMMENTED** [review]
@nicolaspi 

LGTM! Leaves a few minor comments.

> if we constrain `y_true` and `y_pred` to have the same structure

It seems that this PR still requires a single-tensor-in, single-tensor-out signature for the actual loss function. Is it correct?

I encountered issues when implementing OD losses (such as Varifocal Loss) which requires `y_true` to be a tuple of (scores, label) while `y_pred` is logits. Would this use case fit within this PR?

**Comment by nicolaspi** [comment]
> I encountered issues when implementing OD losses (such as Varifocal Loss) which requires `y_true` to be a tuple of (scores, label) while `y_pred` is logits. Would this use case fit within this PR?

The loss requires `y_pred` and `y_true` to have the same pytree structure, but they can be arbitrarily nested. You can implement the Varifocal Loss by changing the model's output with `output={"logits": (logits, logits)}` and configure `loss={'logits': VarifocalLoss}`.

File: keras/src/trainers/compile_utils.py
Comment (nicolaspi) [reply keras/src/trainers/compile_utils.py:489]: Potentially.
This case happens when the loss is applied at the root of the model's output. For example if `output=(y1, {'a':y2, 'b': ...})`. How should the (potentially deeply nested) pytree `(y1, {'a':y2, 'b': ...})` be named?
Current implementation:
```
if root is a dict:
  Uses concatenation of the dictionary keys
else:
  Uses concatenation of `model.output_names`
```

The name is checked for a maximum size, and then append `...` if the size is exceeded.

Do you have suggestion for a naming convention? CC @fchollet

File: keras/src/trainers/compile_utils.py
Comment (nicolaspi) [reply keras/src/trainers/compile_utils.py:412]: It was meant for clarity for developers. `CompileLoss` builds a list `_flat_losses` of those 4-tuples. I made this namedtuple to enhance understanding of what those tuples represent.
What is your suggestion, shall I use regular tuples?

File: keras/src/trainers/compile_utils.py
Comment (nicolaspi) [reply keras/src/trainers/compile_utils.py]: Changed to `elif`

File: keras/src/trainers/compile_utils.py
Comment (james77777778) [reply keras/src/trainers/compile_utils.py:412]: Oh I see. I wasn't aware of the line `CompileLoss.Loss`.
I agree that `namedtuple` is indeed clearer than a regular tuple.

File: keras/src/models/model_test.py
Comment (james77777778) [reply keras/src/models/model_test.py]: This seems a bit complicated to understand and not very intuitive IMO. We might want to document these rules somewhere.
I think we should wait for @fchollet's comment.

Comment (nicolaspi) [reply keras/src/models/model_test.py]: Changed `anti_type` to `other_type` for better clarity.

File: keras/src/trainers/compile_utils.py
Comment (nicolaspi) [reply keras/src/trainers/compile_utils.py]: I removed the truncation.

**Review by fchollet — APPROVED** [review]

**Comment by fchollet** [comment]
The test failure for TF-GPU seems related:

FAILED keras/src/metrics/iou_metrics_test.py::IoUTest::test_compilation - ValueError: `x` and `indices` must have the same number of dimensions, but receive shape [2, 2] and [2].

**Comment by nicolaspi** [comment]
> The test failure for TF-GPU seems related:
> 
> FAILED keras/src/metrics/iou_metrics_test.py::IoUTest::test_compilation - ValueError: `x` and `indices` must have the same number of dimensions, but receive shape [2, 2] and [2].

@fchollet I can't manage to reproduce the issue locally. I tried with the same Python, NVIDIA drivers, and CUDA versions, to no avail.

File: keras/src/models/model_test.py
Comment (nicolaspi) [reply keras/src/models/model_test.py]: @james77777778 can you accept the PR if you believe this is resolved? Thanks!

Comment (james77777778) [reply keras/src/models/model_test.py]: LGTM!

I believe the final decision is up to @fchollet

Comment (nicolaspi) [reply keras/src/models/model_test.py]: Can you approve the changes? The PR shows that it’s missing one reviewer’s approval. Thanks!

Comment (james77777778) [reply keras/src/models/model_test.py]: Unfortunately, I don't have the permission to approve the PR.

Comment (nicolaspi) [reply keras/src/models/model_test.py]: Oh OK! Have a good day!

**Comment by fchollet** [comment]
Thanks for the contribution!

**Comment by fchollet** [comment]
This change appears to break cases like this one:

```python
inputs = keras.Input((2,))
x = keras.layers.Dense(3)(inputs)
model = keras.Model(inputs, [x, x, x])
model.compile(loss=["mse", None, None])

model.fit(x=np.random.random((320, 2)), y=np.random.random((320, 3)))
```

Where we have non-matching structure for model outputs and y_true (but a correct loss configuration)

**Comment by nicolaspi** [comment]
@fchollet Please check this PR: https://github.com/keras-team/keras/pull/20477

**Pull request merged**
